packages/protocol/contracts/common/UniswapFeeHandlerSeller.sol

```diff
@@ -0,0 +1,205 @@
+  }
+
+  /**
+  * @dev Calculates the minimum amount of tokens that can be received for a given amount of sell tokens, 
+          taking into account the slippage and the rates of the sell token and CELO token on the Uniswap V2 pair.
+  * @param sellTokenAddress The address of the sell token.
+  * @param maxSlippage The maximum slippage allowed.
+  * @param amount The amount of sell tokens to be traded.
+  * @param bestRouter The Uniswap V2 router with the best price.
+  * @return The minimum amount of tokens that can be received.
+  */
+  function calculateAllMinAmount(
+    address sellTokenAddress,
+    uint256 maxSlippage,
+    uint256 amount,
+    IUniswapV2RouterMin bestRouter
+  ) private view returns (uint256) {
+    ISortedOracles sortedOracles = getSortedOracles();
+    uint256 minReports = minimumReports[sellTokenAddress];
+
+    require(
+      sortedOracles.numRates(sellTokenAddress) >= minReports,
+      "Number of reports for token not enough"
+    );
+
+    uint256 minimalSortedOracles = 0;
+    // if minimumReports for this token is zero, assume the check is not needed
+    if (minReports > 0) {
+      (uint256 rateNumerator, uint256 rateDenominator) = sortedOracles.medianRate(sellTokenAddress);
+
+      minimalSortedOracles = calculateMinAmount(
+        rateNumerator,
+        rateDenominator,
+        amount,
+        maxSlippage
+      );
+    }
+
+    IERC20 celoToken = getGoldToken();
+    address pair = IUniswapV2FactoryMin(bestRouter.factory()).getPair(
+      sellTokenAddress,
+      address(celoToken)
+    );
+    uint256 minAmountPair = calculateMinAmount(
+      IERC20(sellTokenAddress).balanceOf(pair),
+      celoToken.balanceOf(pair),
+      amount,
+      maxSlippage
+    );
+
+    return Math.max(minAmountPair, minimalSortedOracles);
+  }
+
+  // This function explicitly defines few variables because it was getting error "stack too deep"
+  function sell(
+    address sellTokenAddress,
+    address buyTokenAddress,
+    uint256 amount,
+    uint256 maxSlippage // as fraction,
+  ) external {
+    require(
+      buyTokenAddress == registry.getAddressForOrDie(GOLD_TOKEN_REGISTRY_ID),
+      "Buy token can only be gold token"
+    );
+
+    require(
+      routerAddresses[sellTokenAddress].values.length > 0,
+      "routerAddresses should be non empty"
+    );
+
+    // An improvement to this function would be to allow the user to pass a path as argument
+    // and if it generates a better outcome that the ones enabled that gets used
+    // and the user gets a reward
+
+    IERC20 celoToken = getGoldToken();
+
+    IUniswapV2RouterMin bestRouter;
+    uint256 bestRouterQuote = 0;
+
+    address[] memory path = new address[](2);
+
+    for (uint256 i = 0; i < routerAddresses[sellTokenAddress].values.length; i++) {
+      address poolAddress = routerAddresses[sellTokenAddress].get(i);
+      IUniswapV2RouterMin router = IUniswapV2RouterMin(poolAddress);
+
+      path[0] = sellTokenAddress;
```

Comment:
Do we need to set path in every loop step? Could we move variable assignment outside of the loop ?